src/library/autodj/autodjprocessor.cpp

```diff
@@ -817,61 +843,121 @@ void AutoDJProcessor::calculateTransition(DeckAttributes* pFromDeck,
+    } else if (m_transitionMode == TransitionMode::AlignIntroOutroEnd) {
+        if (outroLength > 0 && introLength > 0) {
+            if (outroLength < introLength) {
+                pFromDeck->fadeBeginPos = outroStart;
+                pFromDeck->fadeDuration = outroLength;
+                pToDeck->startPos = introEnd - outroLength;
+            } else {
+                pFromDeck->fadeBeginPos = outroEnd - introLength;
+                pFromDeck->fadeDuration = introLength;
+                pToDeck->startPos = introStart;
+            }
+        } else if (outroLength > 0 && introLength <= 0) {
+            pFromDeck->fadeBeginPos = outroStart;
+            pFromDeck->fadeDuration = outroLength;
+            pToDeck->startPos = introStart;
+        } else if (outroLength <= 0 && introLength > 0) {
+            pFromDeck->fadeBeginPos = outroEnd - introLength;
+            pFromDeck->fadeDuration = introLength;
+            pToDeck->startPos = introStart;
         } else {
-            pFromDeck->posThreshold = 1.0;
-            pFromDeck->fadeDuration = 0.0;
+            useFixedFadeTime(pFromDeck, pToDeck, outroEnd, introStart);
         }
+    } else if (m_transitionMode == TransitionMode::FixedSkipSilence) {
+        useFixedFadeTime(pFromDeck, pToDeck,
+                         getLastSoundPosition(pFromDeck), getFirstSoundPosition(pToDeck));
+    } else if (m_transitionMode == TransitionMode::FixedLoadAtCue) {
+        useFixedFadeTime(pFromDeck, pToDeck,
+                         getLastSoundPosition(pFromDeck), getMainCuePosition(pToDeck));
+    } else {
+        useFixedFadeTime(pFromDeck, pToDeck, fromTrackDuration, 0);
+    }
 
-        if (toTrackDuration > 0.0) {
-            // Transition time is negative here, so we need to add instead of subtract.
-            pToDeck->startPos = (toTrackIntroStart + m_transitionTime) / toTrackDuration;
-        } else {
-            pToDeck->startPos = 0.0;
-        }
+    // Guard against the next track being too short. This transition must finish
+    // before the next one starts.
+    double toDeckOutroStart = getOutroStartPosition(pToDeck);
+    if (toDeckOutroStart <= 0) {
+        toDeckOutroStart = getOutroEndPosition(pToDeck);
+    }
+    double maxFadeTime = toDeckOutroStart - pToDeck->startPos;
+    if (pFromDeck->fadeDuration > maxFadeTime) {
+        pFromDeck->fadeDuration = maxFadeTime;
+    }
+
+    VERIFY_OR_DEBUG_ASSERT(fromTrackDuration > 0) {
+        pFromDeck->fadeBeginPos = fromTrackDuration;
+        pFromDeck->fadeDuration = 0;
+    }
+    VERIFY_OR_DEBUG_ASSERT(toTrackDuration > 0) {
+        pFromDeck->fadeDuration = 0;
+        pToDeck->startPos = 0;
+    }
+
+    // These are expected to be a fraction of the track length.
+    pFromDeck->fadeBeginPos /= fromTrackDuration;
+    pFromDeck->fadeDuration /= fromTrackDuration;
+    pToDeck->startPos /= toTrackDuration;
+}
+
+void AutoDJProcessor::useFixedFadeTime(DeckAttributes* pFromDeck, DeckAttributes* pToDeck,
```

Comment:
You're right, "fixed" is not the best term to use for this... but I'm not sure what to replace it with...

Reply:
How about renaming the function to "useSpinboxTransitionTime"?

Reply:
Great. However this is too long for the transition timing combo box. So we need something else there.

Reply:
Yeah I have though the "Transition time" label for the combobox in the AutoDJ GUI is now misleading... but I'm not sure how else to label it because its function is different depending on the transition mode.